File: src/com/leetcode/Medium/WordBreak.java
package com.leetcode.Medium;

import java.util.*;

// 139. Word Break
public class WordBreak {
    public boolean wordBreak(String s, List<String> wordDict) {
        Queue<Integer> queue = new LinkedList<>();
        Set<String> wordDictSet = new HashSet<>(wordDict);
        boolean[] visited = new boolean[s.length()];
        queue.add(0);

        while (!queue.isEmpty()) {
            int i = queue.poll();
            if (visited[i]) continue;

            for (int j = i + 1; j <= s.length(); j++) {
                if (wordDictSet.contains(s.substring(i, j))) {
                    if (j == s.length())
                        return true;

                    queue.add(j);
                }
            }
            visited[i] = true;
        }

        return false;
    }
}

File: JavaScript/Medium/WordBreak.js
/**
 * 139. Word Break
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
  let queue = [];
  let wordDictSet = new Set(wordDict);
  let visited = new Array(s.length).fill(false);
  queue.push(0);

  while (queue.length !== 0) {
    let i = queue.shift();

    if (visited[i])
      continue;

    for (let j = i + 1; j <= s.length; j++) {
      if (wordDictSet.has(s.substring(i, j))) {
        if (j === s.length)
          return true;

        queue.push(j);
      }
    }
    visited[i] = true;
  }
  return false;
};